### ISSUE TITLE
Nested hybrid_adapter should unwrap

### ISSUE THREAD

**Issue by Morwenn** [Morwenn/cpp-sort#6]
Currently, because of the way `hybrid_adapter` and `sorter_traits` work, the following does not pick the most suitable algorithm possible:

```
using sorter = cppsort::hybrid_adapter<
    cppsort::hybrid_adapter<
        forward_sorter,
        random_access_sorter
    >,
    bidirectional_sorter
>;
```

If we feed a random-access iterable to `sorter`, it will pick `bidirectional_sorter` instead of `random_access_sorter` because the iterator category of the nested `hybrid_adapter` is `forward_sorter`. To overcome this problem, `hybrid_adapter` should unwrap so that the sorter above is equivalent to:

```
using sorter = cppsort::hybrid_adapter<
    forward_sorter,
    random_access_sorter
    bidirectional_sorter
>;
```

Unfortunately, this is not _trivial_ since `hybrid_adapter` already performs some heavy wizardry and even a bit of back magic. Basically, it would be nice to be able to write the following specialization:

```
template<typename... TT, typename... UU, typename... VV>
struct hybrid_adapter<TT..., hybrid_adapter<UU...>, VV...>:
    hybrid_adapter<TT..., UU..., VV...>
{};
```

Unfortunately, it's not possible to expand a parameter pack on the left like that so it needs to be done differently. Implementing that proved to be complex and I still did not manage to tackle the problem, be it from the inside of `hybrid_adapter` or from the outside. The closest thing I got is this:

```
template<typename... Sorters1, typename... Sorters2>
struct hybrid_adapter<hybrid_adapter<Sorters1...>, Sorters2...>:
    hybrid_adapter<Sorters1..., Sorters2...>
{};
```

Unfortunately, it only unwraps nested `hybrid_adapter` if it appears at the beginning of the parameter pack, so it only solves half of the problem. Having a half-fix is worse than no fix at all in this case though, so I won't implement it in master.

**Comment by Morwenn** [comment]
While not solving the unwrapping problem, 3863e22fbb4a93f2175cd530a6d3c933a84c5c78 improves the support for nested `hybrid_adapter`; it makes the following code valid as expected:

    using sorter = hybrid_adapter<
        spread_sorter,
        default_sorter
    >;

This modification should still play well with the eventual future unwrapping of nested `hybrid_adapter`

**Comment by Morwenn** [comment]
Closed by #140, thanks @foundry27 :)
